### PR TITLE
BPE ASR Datalayer

### PR DESCRIPTION
**Pull request by blisc**
Usage example:
```python
tokenizer = nemo_nlp.data.NemoBertTokenizer(pretrained_model="bert-base-uncased")
data = nemo_asr.data_layer.AudioToTextBPEDataLayer(
    manifest_filepath=args.train_dataset,
    batch_size=args.batch_size,
    tokenizer=tokenizer,
    **asr_params['AudioToTextDataLayer']['train'],
)
```

### CODE REVIEW

File: nemo/collections/asr/parts/dataset.py [hunk 1]
```diff
@@ -228,6 +368,173 @@ def __len__(self):
...
+
+        t, tl = sample.text_tokens, len(sample.text_tokens)
+
+        return f, fl, torch.tensor(t).long(), torch.tensor(tl).long()
+
+    def __len__(self):
+        return len(self.collection)
+
+
+class AudioCharDataset(_AudioDataset):
```
Comment (titu1994) [inline nemo/collections/asr/parts/dataset.py:453]: If `AudioDataset` is deprecated post v0.12, should we have backward compat for older scripts by aliasing `AudioDataset = AudioCharDataset`?

File: nemo/collections/asr/data_layer.py [hunk 2]
```diff
@@ -358,6 +214,321 @@ def data_iterator(self):
...
+    @property
+    def dataset(self):
+        return None
+
+    @property
+    def data_iterator(self):
+        return self._dataloader
+
+
+class AudioToTextCharDataLayer(_AudioToTextDataLayer):
```
Comment (titu1994) [inline nemo/collections/asr/data_layer.py:336]: If `AudioToTextDataLayer` is being deprecated in v0.12, we should probably alias `AudioToTextDataLayer = AudioToTextCharDataLayer` to maintain compat with older training scripts

**Comment by tkornuta-nvidia** [comment]
Hey, just one comment: **NemoBertTokenizer**

I am against adding Nemo in the name of the modules, please - if we would like to be consistent, then **every class in NeMo** should have this prefix, what is the point?

**Comment by blisc** [comment]
> Hey, just one comment: **NemoBertTokenizer**
> 
> I am against adding Nemo in the name of the modules, please - if we would like to be consistent, then **every class in NeMo** should have this prefix, what is the point?

That is something you have to ask the nlp team. Note, this changes makes nemo_asr depends on nemo_nlp for the tokenizer code.

**Comment by blisc** [comment]
Closing in favour of #835

**Pull request closed without merging**